2_process/src/add_dynamic_attr.R

```diff
@@ -0,0 +1,796 @@
+                                     baseflow = NULL, 
+                                     CAT_Land = NULL, Upstream_Land = NULL, 
+                                     gridMET = NULL,
+                                     attr_prefix, Upstream_Land_prefix = NULL,
+                                     lag_table){
+  #' @description computes dynamic attributes for each reach based on the provided dates
+  #' 
+  #' @param attrs table of static attributes (columns) for each reach (rows)
+  #' @param dyn_cols character vector of keywords to search for columns that 
+  #' should be converted into dynamic attributes.
+  #' @param start_date first date to compute attributes
+  #' @param end_date last date to compute attributes
+  #' @param baseflow table of monthly baseflow estimates for each reach
+  #' @param CAT_Land table of catchment land cover class proportions
+  #' @param Upstream_Land table of total upstream land cover class proportions
+  #' @param gridMET table of gridment data. Row length = dates x PRMS reach
+  #' @param attr_prefix character vector of prefixes for the attributes 
+  #' (e.g., CAT, TOT)
+  #' @param Upstream_Land_prefix character vector of prefixes for the upstream 
+  #' land cover attributes (e.g., TOT, ACC)
+  #' @param lag_table table with 3 columns: attributes, lags, and lag_unit.
+  #' This is used to compute the lagged attributes
+  #' 
+  #' @return tbl with the added features.
+  
+  #make a dynamic features dataframe
+  dates <- seq(as.Date(start_date), as.Date(end_date), 1)
+  segs <- sort(unique(attrs$PRMS_segid))
+  df <- tibble(Date = dates, seg = segs[1])
+  for(i in 2:length(segs)){
+    df_tmp <- tibble(Date = dates, seg = segs[i])
+    df <- rbind(df, df_tmp)
+  }
+  rm(df_tmp, dates, i)
+  
+  #Special handling for each variable
+  if ('HDENS' %in% dyn_cols){
+    #Housing density has the format HDENSYY
+    tmp_attrs <- select(attrs, PRMS_segid, contains('HDENS'))
+    
+    #Convert the column names to 4-digit years
+    col_years <- str_split(colnames(tmp_attrs)[-1], 
+                           pattern = '_', simplify = TRUE)[,2] %>%
+      substr(start = 6, stop = 7) %>%
+      get_four_digit_year()
+    tmp_colnames <- colnames(tmp_attrs)
+    for(i in 2:length(tmp_colnames)){
+      tmp_colnames[i] <- gsub(x = tmp_colnames[i], pattern = 'HDENS+[0-9]+[0-9]', 
+                              replacement = col_years[i-1]) %>%
+        #Format column names to match requirements for the function
+        gsub(pattern = '_area_wtd', replacement = '')
+    }
+    colnames(tmp_attrs) <- tmp_colnames
+    rm(i, tmp_colnames, col_years)
+    
+    #HDENS data date ranges and dataset years
+    tmp_date_ranges <- seq(as.Date('1965-09-30'), as.Date('2005-09-30'),
+                           '10 years')
+    tmp_years <- tmp_attrs %>% 
+      select(starts_with(attr_prefix[1])) %>% 
+      colnames() %>% 
+      sort() %>% 
+      str_split(pattern = '_', simplify = TRUE) %>% 
+      .[,2]
```

Comment:
Similar to my comment above, I think extracting the digits rather than specific indices within the `str_split` output is a little easier to read and interpret.

```r
tmp_years <- tmp_attrs %>% 
      select(starts_with(attr_prefix[1])) %>% 
      colnames() %>% 
      stringr::str_extract("[[:digit:]]+") %>%
      sort()
```

Reply:
I'll edit this for all of the attributes